src/apis/middleware.rs

```diff
@@ -4,30 +4,166 @@ use reqwest::{IntoUrl, Method};
+            retry_backoff_factor: Option<f32>,
+            retry_backoff_jitter: Option<f32>,
+            retry_backoff_max: Option<f32>,
+        }
+
+        let helper = Helper::deserialize(deserializer)?;
+        let default = BackoffParams::default();
+
+        Ok(BackoffParams {
+            max_retries: helper.max_retries.unwrap_or(default.max_retries),
+            retry_backoff_factor: helper
+                .retry_backoff_factor
+                .unwrap_or(default.retry_backoff_factor),
+            retry_backoff_jitter: helper
+                .retry_backoff_jitter
+                .unwrap_or(default.retry_backoff_jitter),
+            retry_backoff_max: helper
+                .retry_backoff_max
+                .unwrap_or(default.retry_backoff_max),
+        })
+    }
+}
+
+impl Default for BackoffParams {
+    fn default() -> Self {
+        Self {
+            max_retries: 3,
+            retry_backoff_factor: 1.0,
+            retry_backoff_jitter: 3.0,
+            retry_backoff_max: 60.0,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LimiterParams {
+    pub max_requests: usize,
+    pub window: Duration,
+}
+
+impl<'de> Deserialize<'de> for LimiterParams {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct Helper {
+            max_requests: Option<usize>,
+            window: Option<Duration>,
+        }
+
+        let helper = Helper::deserialize(deserializer)?;
+        let default = LimiterParams::default();
+
+        Ok(LimiterParams {
+            max_requests: helper.max_requests.unwrap_or(default.max_requests),
+            window: helper.window.unwrap_or(default.window),
+        })
+    }
+}
+
+impl Default for LimiterParams {
+    fn default() -> Self {
+        Self {
+            max_requests: 5,
+            window: Duration::from_secs(1),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+struct Limiter {
+    params: LimiterParams,
+    requests: VecDeque<Instant>,
+}
+
+impl Limiter {
+    pub fn new(params: LimiterParams) -> Self {
+        Self {
+            params,
+            requests: VecDeque::new(),
+        }
+    }
+
+    pub fn acquire(&mut self) {
+        let now = Instant::now();
+        self.cleanup_old_requests(now);
+
+        // If we're at the limit, wait until we can make another request
+        if self.requests.len() >= self.params.max_requests {
+            if let Some(oldest) = self.requests.front() {
+                let wait_time = self
+                    .params
+                    .window
+                    .saturating_sub(now.duration_since(*oldest));
+                if !wait_time.is_zero() {
+                    std::thread::sleep(wait_time);
+                }
+
+                // Clean up again after sleeping
+                let now = Instant::now();
+                self.cleanup_old_requests(now);
```

Comment:
Why do you need to do a cleanup again ? The next request will start by a cleanup...